### PR TITLE
Wait for APIService status when installing karmada with `karmadactl init`

### LINKED ISSUE #1422: Add check for cluster apiservice when install karmada with kubectl-karmada
**What would you like to be added**:

Add ready check for `v1alpha1.cluster.karmada.io` apiservice when install karmada with `kubectl-karmada`, as discussed in https://github.com/karmada-io/karmada/pull/1306#issuecomment-1055167142

**Why is this needed**:

Without ready check, if the AVAILABLE of `v1alpha1.cluster.karmada.io` apiservice is False, cluster join will failed.

### PR DESCRIPTION
**Pull request by zgfh**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it**:
Add check for cluster apiservice when install karmada with kubectl-karmada
**Which issue(s) this PR fixes**:
Fixes #1422

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
feature: Add check for cluster apiservice when install karmada with kubectl-karmada
```

### CODE REVIEW

**Comment by prodanlabs** [comment]
looks good

/lgtm
/approve

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm

File: pkg/karmadactl/cmdinit/karmada/check.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	aggregator "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+)
+
+func waitAPIServiceReady(c *aggregator.Clientset, name string, timeout int) error {
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/karmada/check.go:14]:
```suggestion
func waitAPIServiceReady(c *aggregator.Clientset, name string, timeout time.Duration) error {
```
Why not use the `time.Duration`?

File: pkg/karmadactl/cmdinit/karmada/check.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	if err := wait.PollImmediate(time.Second, time.Duration(timeout)*time.Second, func() (done bool, err error) {
+		apiService, e := c.ApiregistrationV1().APIServices().Get(context.TODO(), name, metav1.GetOptions{})
+		if e != nil {
+			return false, nil
+		}
+		for _, condition := range apiService.Status.Conditions {
+			if condition.Type == apiregistrationv1.Available {
+				if condition.Status == apiregistrationv1.ConditionTrue {
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/karmada/check.go:22]: Can we use the [IsAPIServiceConditionTrue](https://github.com/kubernetes/kubernetes/blob/52cd4d53ac8ac33cacd3d2c618cf22f1e3fd6a7e/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/helper/helpers.go#L125)?

File: pkg/karmadactl/cmdinit/karmada/check.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+					klog.Infof("apiservice with name %s desired condition:  %v = %v, currently: %v", name, apiregistrationv1.Available, apiregistrationv1.ConditionTrue, condition.Status)
+					return true, nil
+				}
+				klog.Infof("apiservice with name %s desired condition:  %v = %v, currently: %v", name, apiregistrationv1.Available, apiregistrationv1.ConditionTrue, condition.Status)
+				return false, nil
+			}
+		}
+
+		klog.Infof("apiservice with name %s desired condition:  %v = %v, currently: ... ", name, apiregistrationv1.Available, apiregistrationv1.ConditionTrue)
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/karmada/check.go:31]:
```suggestion
		klog.Infof("Waiting for APIService(%s) condition(%s), will try", name, apiregistrationv1.Available)
```

**Review by RainbowMango — COMMENTED** [review]
Only some nits. Otherwise LGTM.

File: pkg/karmadactl/cmdinit/karmada/check.go (hunk 1 above)
Comment (zgfh) [reply pkg/karmadactl/cmdinit/karmada/check.go:14]: copy from here:  https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/cmdinit/kubernetes/check.go#L91

Is there any specification for this? If there is, I am happy to modify them together( i am golang beginer)

File: pkg/karmadactl/cmdinit/karmada/deploy.go [hunk 4]
```diff
@@ -231,6 +232,8 @@ func initAPIService(clientSet *kubernetes.Clientset, restConfig *rest.Config) er
 	if _, err := apiRegistrationClient.ApiregistrationV1().APIServices().Create(context.TODO(), aaAPIService, metav1.CreateOptions{}); err != nil {
 		return err
 	}
-
+	if err := waitAPIServiceReady(apiRegistrationClient, aaAPIServiceObjName, time.Duration(120)*time.Second); err != nil {
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/karmada/deploy.go:235]:
```suggestion
	if err := waitAPIServiceReady(apiRegistrationClient, aaAPIServiceObjName, 120*time.Second); err != nil {
```
Redundant conversion here.

File: pkg/karmadactl/cmdinit/karmada/check.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	helper "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/helper"
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/karmada/check.go:11]:
```suggestion
	apiregistrationv1helper "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/helper"
```

**Review by RainbowMango — COMMENTED** [review]
hack/../vendor, 'go.mod' or 'go.sum' is out of date. Please run hack/update-vendor.sh

**Comment by RainbowMango** [comment]
/retitle Wait for APIService status when installing karmada with `karmadactl init`

**Comment by zgfh** [comment]
@RainbowMango thanks, updated

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

Thanks @zgfh !

**Pull request merged**